Various fixes for IPv4 autoconf

man-gc opened the pull request:
- Fixes allowed address range
- Do not set address as assigned until conflict detection is done
- Introduce locking to avoid race conditions when changing state

man-gc committed:
net: ipv4_autoconf: fix address range

IANA allocated the 169.254/16 address range for IPv4 Link-Local
addresses but RFC3927 reserved the first and last 256 adresses for
future use. Therefore, they should not be assigned to any interface.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
net: ipv4_autoconf: introduce ALLOCATING state

Do not consider address assigned until conflict detection is finished,
so introduce a new state to handle the case where an address is reserved
but not should not be considered usable until probe is finished.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

rlubos reviewed (commented):
Address range and mutex fixes look fine, but the remaining changes are unclear to me and the commit messages don't really explain **why**.

man-gc commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -81,17 +76,30 @@ static void acd_event_handler(struct net_mgmt_event_callback *cb,
+		goto out;
+	}
+
+	if (cfg->ipv4auto.state != NET_IPV4_AUTOCONF_ALLOCATING) {
```
OK, removing the line to implement the alternative (a) from RFC for now.

man-gc commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -144,9 +167,29 @@ void net_ipv4_autoconf_reset(struct net_if *iface)
 
+	(void) k_mutex_lock(&lock, K_FOREVER);
+
+	switch (cfg->ipv4auto.state) {
```
Using a lock, I could not call `reset()` inside `start()` with lock held.
The rest is only to ensure a coherent state after calling the method. Maybe it should be split in its own commit.

man-gc committed:
net: ipv4_autoconf: force state after start/reset calls

Ensure the state is either INIT or RENEW after calling start()/reset().
Other states are only valid during active management of the link-local
IPv4 address of the interface.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
net: ipv4_autoconf: introduce locking

Until now, modifications to the ipv4auto configuration has been handled
lock-free but this can lead to race conditions between the configuration
and conflict detection threads.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>
